Add a converter from mamba_ssm -> huggingface mamba

byi8220 opened the pull request:
# What does this PR do?
Fixes #29631

This PR creates a new command line script `convert_mamba_ssm_checkpoint_to_pytorch.py` which converts model checkpoints created by the [state-spaces/mamba](https://github.com/state-spaces/mamba/tree/main) repo into a Huggingface `MambaForCausalLM` model.

The intended usage of this script is:
```
python src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py \
--mamba_checkpoint_file=path/to/pytorch_model.bin \
--config_json_file=/path/to/ssm_config.json \
--output_dir=/path/to/out/dir
```

This script has a dependency on the `mamba_ssm` package.

### Testing

A validation pass is performed before exporting the model.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

byi8220 committed:
implement convert_mamba_ssm_checkpoint_to_pytorch

byi8220 committed:
Add test test_model_from_mamba_ssm_conversion

byi8220 committed:
moved convert_ssm_config_to_hf_config to inside mamba_ssm_available check

byi8220 committed:
fix skipif clause

byi8220 committed:
moved skips to inside test since skipif decorator isn't working for some reason

amyeroberts commented:
Thanks for opening this PR @byi8220 ! Let us know when it's ready for review 🤗

byi8220 commented:
Thanks, I think it should be reviewable now?

amyeroberts commented on tests/models/mamba/test_modeling_mamba.py:
```diff
@@ -489,3 +500,49 @@ def test_simple_generate_cuda_kernels_big(self, device):
 
         self.assertEqual(output_sentence, expected_output)
+
+    @parameterized.expand([("state-spaces/mamba-130m",)])
```
I don't think it's necessary to add a test - we don't do this for other models. Instead we typically compare the original model's logits to our converted model's for a given output in the conversion script

byi8220 replied:
Makes sense. Removed the test and ran the command e2e and ran a spot test (https://gist.github.com/byi8220/9b44a5f6c6c2c7533704801478f1760a)

I was able to successfully convert their base 130m and 790m models without a problem, however my GPU setup runs out of memory trying for the 1.4b and 2.8b models.

amyeroberts reviewed (commented):
Thanks for working on this!

Just a comment about having validation in the script rather than a new test

byi8220 committed:
Merge branch 'huggingface:main' into mamba-converter

byi8220 committed:
Added validation

byi8220 committed:
removed test

byi8220 committed:
Merge branch 'mamba-converter' of https://github.com/byi8220/transformers into mamba-converter

byi8220 committed:
fixup

byi8220 committed:
only compare logits

byi8220 committed:
Merge branch 'huggingface:main' into mamba-converter

byi8220 committed:
Merge branch 'huggingface:main' into mamba-converter

byi8220 committed:
remove weight rename

byi8220 committed:
Merge branch 'huggingface:main' into mamba-converter

amyeroberts commented on src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,153 @@
+
+
+if is_mamba_ssm_available():
+    from mamba_ssm.models.config_mamba import MambaConfig as MambaConfig_ssm
```
nit - camel casing for classes

```suggestion
    from mamba_ssm.models.config_mamba import MambaConfig as MambaConfigSSM
```

byi8220 replied:
Done

amyeroberts commented on src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,153 @@
+    original_state_dict: dict, original_ssm_config_dict: dict, hf_model: MambaForCausalLM, tokenizer: AutoTokenizer
+) -> None:
+    """Validate the converted model returns the same output as the original model."""
+    if not is_mamba_ssm_available():
+        raise ImportError(
+            "Calling validate_converted_model requires the mamba_ssm library to be installed. Please install it with `pip install mamba_ssm`."
+        )
+    if not torch.cuda.is_available():
+        raise ValueError(
+            "This script is to be run with a CUDA device, as the original mamba_ssm model does not support cpu."
+        )
```
Let's move these into `convert_mamba_checkpoint_file_to_huggingface_model_file`. 

Each function doesn't need to handle imports, we can delegate this to a single "function" which calls others and is responsible for input / env validation

byi8220 replied:
Done

amyeroberts reviewed (approved):
Thanks for adding this! 

Looks good to me, just some small nits. cc @ArthurZucker for reference

byi8220 committed:
Update src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

byi8220 committed:
nits

byi8220 committed:
Merge branch 'huggingface:main' into mamba-converter

byi8220 commented:
Addressed nits.

Since there's no unit test for this anymore just did one last sanity test. Works on my machine

1. My script used to spot test the conversion (https://gist.github.com/byi8220/9b44a5f6c6c2c7533704801478f1760a) passes for the 130m and 790m models. My GPU doesn't have enough memory to run the larger ones.
2. Downloading the weights and config of a [130m model online](https://huggingface.co/state-spaces/mamba-130m/tree/main) and running the command `python src/transformers/models/mamba/convert_mamba_ssm_checkpoint_to_pytorch.py --mamba_checkpoint_file=/tmp/scratch/pytorch_model.bin --config_json_file=/tmp/scratch/config.json --output_dir=/tmp/scratch/out` appears to function e2e.

amyeroberts commented:
Awesome work - thanks again for adding this and for running some sanity checks!
